### ISSUE TITLE
Write information to the timeline when an object that has a `::dispose` method is destructed without `dispose` being called.

### ISSUE THREAD

**Issue by dnfield** [flutter/flutter#83274]
Many objects in `dart:ui` have a `dispose` method that eagerly releases native resources when called.

It is almost always a mistake to not call `dispose` on an object before it is GC'd. We should write information about this to the timeline, and then devtools could be enhanced to let the developer know when/how often this happened.

/cc @jacob314 @terrylucas 
/cc @jason-simmons

**Comment by dnfield** [comment]
For some objects, like `Picture`s, we could also add information that would allow the user to view what picture it was.

**Comment by dnfield** [comment]
Part of https://github.com/flutter/flutter/issues/81514

**Comment by dnfield** [comment]
One question I'm not sure on is whether we should also/alternatively expose this to a different service endpoint that could be used more efficiently than scanning the entire timeline.

**Comment by jacob314** [comment]
Why do I care exactly when the object was GCed without being disposed? I would have expected I would care about which objects were GCed without being first disposed so that I could fix the logic bug causing the issue.

Once Dart supports finalization tracking from Dart code, you could surface this information from Dart code rather than depending on C++ code in the engine.
https://github.com/dart-lang/sdk/issues/45455

**Comment by dnfield** [comment]
+1 on Dart exposing this - if we're confident that's going forward, we could probably skip this. But if that's going to take a while, we can probably implement this a lot more quickly for `dart:ui` objects.

I'm a bit confused here: 

> Why do I care exactly when the object was GCed without being disposed? I would have expected I would care about which objects were GCed without being first disposed so that I could fix the logic bug causing the issue.

Did you mistype something or am I misreading it? :)

The idea here is that if a GC happened before `dispose` was called, you ended up relying on a GC to dispose of native resources held by that object. This is bad, because it means you held native resources longer than you should have. This is probably normally an error in the framework, but could also be introduced by user code, e.g. if you directly created a `Picture` object and never called `dispose` on it.

**Comment by dnfield** [comment]
E.g. this program should trigger the warning(s):

```dart
import 'dart:ui';

void main() {
  window.onBeginFrame = (_) {
    final SceneBuilder builder = SceneBuilder();
    final Scene scene = builder.build();
    window.render(scene);
    // oops, didn't dispose scene
  };
  window.scheduleFrame();
}
```

**Comment by jason-simmons** [comment]
One option that might be worth looking into is `Dart_FinalizableHandle`/`Dart_WeakPersistentHandle`.  These APIs allow native code to set up finalizer callbacks that are invoked after an object is GCed.

Using these APIs, it should be possible to attach a finalizer callback and a tracking object to the engine objects that have `dispose` methods.  If the finalizer is invoked but the tracking object indicates that `dispose` was never called, then the callback could log a description of the object to the console or the timeline.

**Comment by dnfield** [comment]
I think we can just do this using the `DartWrappable` interface, e.g. here (using a log instead of a trace).

We also may want to consider letting objects opt out of this. Some objects don't implement a `dispose` method, and probably don't need to - e.g. `Path` objects are unlikely to benefit much here.

Here's a sample patch that uses `DartWrappable`: https://github.com/flutter/engine/pull/26375

**Comment by jacob314** [comment]
> +1 on Dart exposing this - if we're confident that's going forward, we could probably skip this. But if that's going to take a while, we can probably implement this a lot more quickly for `dart:ui` objects.

I expect dart-lang/sdk#45455 will land within a quarter. 

> 
> I'm a bit confused here:
> 
> > Why do I care exactly when the object was GCed without being disposed? I would have expected I would care about which objects were GCed without being first disposed so that I could fix the logic bug causing the issue.
> 
> Did you mistype something or am I misreading it? :)
> 
> The idea here is that if a GC happened before `dispose` was called, you ended up relying on a GC to dispose of native resources held by that object. This is bad, because it means you held native resources longer than you should have. This is probably normally an error in the framework, but could also be introduced by user code, e.g. if you directly created a `Picture` object and never called `dispose` on it.

I just mean that you don't care "when", you care about the "what". For example, if you dispatched a timeline event, the timeline even would have. time that doesn't seem meaningful. It just indicates what time a GC occurred that caught the issue. That may be multiple frames after the relevant frame with the issue so showing it on the timeline would be confusing. More relevant data would be the "what". For example, the SceneBuilder created on line five of main.dart created at timestamp FOO was GCed before it disposed.

**Comment by dnfield** [comment]
I agree that the when is less important than the what. I'd be completely open to another mechanism to report/record this, but am hesitant to add a new one unnecessarily.

We could certainly add information about what source line created the object, although I'm not sure if we could do that efficiently...

**Comment by zanderso** [comment]
In a debug build you can attach the Dart call stack at object creation to the object, and log it or add it to the timeline event arguments in the native finalizer if `dispose` was not called.

**Comment by jacob314** [comment]
There is already. feature in the memory page of devtools to track object allocations for Dart objects. You have to specify which classes you wish to track. It would be great to get a list of ll classes from the engine that are useful to track by default in debug builds when users are debugging memory issues. We know the identity hash going along with each allocation so as long as you log the identity hash as part of the event indicating an undisposed object, we can join the data sources together.

**Comment by dnfield** [comment]
For a picture we could even serialize the picture if it wasn't disposed, and it could be re-rendered, although I'm not sure if that would include things like images and fonts...

**Comment by dnfield** [comment]
For images we definitely could.

**Comment by jacob314** [comment]
Serializing the picture is an interesting idea. That probably would require integrating with Dart_FinalizableHandle  rather than using FinalizationRegistry as the Dart object and probably the native handle have both probably been destroyed by the time FinalizationRegistry is called.

**Comment by dnfield** [comment]
I think we have some agreement that the timeline is not the right mechanism for this. I'm exploring whether there's some other interface we can use, e.g. `postEvent` or some new callback on `window`/`platformDelegate`.

**Comment by zathras** [comment]
> Why do I care exactly when the object was GCed without being disposed? I would have expected I would care about which objects were GCed without being first disposed so that I could fix the logic bug causing the issue.
> 
> Once Dart supports finalization tracking from Dart code, you could surface this information from Dart code rather than depending on C++ code in the engine.
> [dart-lang/sdk#45455](https://github.com/dart-lang/sdk/issues/45455)

To answer this question:  You care when the native heap runs out of space, because the GC hasn't gotten around to GCing an object that holds substantial native resources.

This kind of thing can and does cause crashes.  I've seen this problem going back to JDK 1.0.2, on pJava, and then again on Android.  Yep, Android faithfully copied the bug that had long been resolved in Java, when they re-implemented the API.  In the case of JDK and Android, we eventually ended up being forced to strictly avoid native allocation whenever possible, because the equivalent of dispose() (on `java.awt.Image`, in our case) wasn't sufficiently encouraged, or normatively required.  On devices like Blu-ray, we added normative requirements around the dispose() equivalent.

Avoiding malloc() is a painful thing to do.  It's probably not too late to avoid that eventual fate in Flutter, if sufficient normative language around dispose is added, and if tooling support like what is described here is created.

**Comment by polina-c** [comment]
Work in progress on this: https://github.com/flutter/devtools/issues/3951
